okhttp/src/test/java/okhttp3/CallTest.kt

```diff
@@ -0,0 +1,4168 @@
+  }
+
+  @Test
+  fun postBodyRetransmittedAfterAuthorizationFail_HTTPS() {
+    enableTls()
+    postBodyRetransmittedAfterAuthorizationFail("abc")
+  }
+
+  @Test
+  fun postBodyRetransmittedAfterAuthorizationFail_HTTP_2() {
+    enableProtocol(Protocol.HTTP_2)
+    postBodyRetransmittedAfterAuthorizationFail("abc")
+  }
+
+  /** Don't explode when resending an empty post. https://github.com/square/okhttp/issues/1131  */
+  @Test
+  fun postEmptyBodyRetransmittedAfterAuthorizationFail() {
+    postBodyRetransmittedAfterAuthorizationFail("")
+  }
+
+  @Test
+  fun postEmptyBodyRetransmittedAfterAuthorizationFail_HTTPS() {
+    enableTls()
+    postBodyRetransmittedAfterAuthorizationFail("")
+  }
+
+  @Test
+  fun postEmptyBodyRetransmittedAfterAuthorizationFail_HTTP_2() {
+    enableProtocol(Protocol.HTTP_2)
+    postBodyRetransmittedAfterAuthorizationFail("")
+  }
+
+  @Throws(Exception::class)
+  private fun postBodyRetransmittedAfterAuthorizationFail(body: String) {
+    server.enqueue(MockResponse().setResponseCode(401))
+    server.enqueue(MockResponse())
+
+    val request = Request.Builder()
+        .url(server.url("/"))
+        .method("POST", body.toRequestBody(null))
+        .build()
+
+    val credential = Credentials.basic("jesse", "secret")
+    client = client.newBuilder()
+        .authenticator(RecordingOkAuthenticator(credential, null))
+        .build()
+
+    val response = client.newCall(request).execute()
+    assertThat(response.code).isEqualTo(200)
+    response.body!!.close()
+
+    val recordedRequest1 = server.takeRequest()
+    assertThat(recordedRequest1.method).isEqualTo("POST")
+    assertThat(recordedRequest1.body.readUtf8()).isEqualTo(body)
+    assertThat(recordedRequest1.getHeader("Authorization")).isNull()
+
+    val recordedRequest2 = server.takeRequest()
+    assertThat(recordedRequest2.method).isEqualTo("POST")
+    assertThat(recordedRequest2.body.readUtf8()).isEqualTo(body)
+    assertThat(recordedRequest2.getHeader("Authorization")).isEqualTo(credential)
+  }
+
+  @Test
+  fun attemptAuthorization20Times() {
+    for (i in 0..19) {
+      server.enqueue(MockResponse().setResponseCode(401))
+    }
+    server.enqueue(MockResponse().setBody("Success!"))
+
+    val credential = Credentials.basic("jesse", "secret")
+    client = client.newBuilder()
+        .authenticator(RecordingOkAuthenticator(credential, null))
+        .build()
+
+    executeSynchronously("/")
+        .assertCode(200)
+        .assertBody("Success!")
+  }
+
+  @Test
+  fun doesNotAttemptAuthorization21Times() {
+    for (i in 0..20) {
+      server.enqueue(MockResponse().setResponseCode(401))
+    }
+
+    val credential = Credentials.basic("jesse", "secret")
+    client = client.newBuilder()
+        .authenticator(RecordingOkAuthenticator(credential, null))
+        .build()
+
+    try {
+      client.newCall(Request.Builder().url(server.url("/0")).build()).execute()
+      fail()
```

Comment:
I'm tempted to introduce an executeFailing() extension method to avoid this repeated pattern, but the current extra line is pretty explanatory, so avoided for now.

Reply:
There's something in the Kotlin test artifact that does this.